horovod/mxnet/mpi_ops.cc

```diff
@@ -109,10 +175,56 @@ void DoHorovodOperation(void*, void* on_complete_ptr, void* param) {
     }
     hvd_contexts.push_back(ctx);
     callbacks.emplace_back([on_complete, ops_param, callback_mutex, i](const Status& status) {
+      auto input_tensor = ops_param->input_tensors[i].get();
+      auto output_tensor = ops_param->output_tensors[i].get();
 #if HAVE_CUDA
       auto hvd_event = status.event;
       if (hvd_event.event) {
+#if MXNET_ASYNC_GPU_ENGINE_SUPPORTED
+        auto async_engine_enabled = dmlc::GetEnv("MXNET_ASYNC_GPU_ENGINE", false);
+        if (async_engine_enabled) {
+          {
+            auto &sync_obj = input_tensor->var()->sync_object;
+            std::lock_guard<std::mutex> l(sync_obj.mutex);
+            // If some reader event is already recorded on the same stream,
+            // we want to replace ourselves by it
+            int i;
+            for (i = 0; i < sync_obj.reader_events.size(); ++i) {
+              auto stream = sync_obj.reader_events[i].stream;
+              if (stream == hvd_event.stream) {
+                sync_obj.reader_events[i].event = hvd_event.event;
+                sync_obj.reader_events[i].pool_index = hvd_event.event_idx;
+                break;
+              }
+            }
+            if (i == sync_obj.reader_events.size()) {
+              sync_obj.reader_events.push_back({hvd_event.event, hvd_event.stream, hvd_event.event_idx});
+            }
+          }
+
+          {
+            auto &sync_obj = output_tensor->var()->sync_object;
+            std::lock_guard<std::mutex> l(sync_obj.mutex);
+            sync_obj.reader_events.clear();
+            sync_obj.writer_event.clear();
+            sync_obj.writer_event.push_back({hvd_event.event, hvd_event.stream, hvd_event.event_idx});
+          }
+
+          if (ops_param->received_splits_tensor) {
+            {
```

Comment:
Nit: Is this scope needed?

Reply:
Good catch, I'll remove it when I rebase for test fixes.